controllers/spacerequest/spacerequest_controller.go

```diff
@@ -486,7 +487,18 @@ func (r *Reconciler) ensureSecretForProvisionedNamespaces(ctx context.Context, m
 
 		case len(secretList.Items) > 1:
 			// some unexpected issue causing to many secrets
-			return fmt.Errorf("invalid number of secrets found. actual %d, expected %d", len(secretList.Items), 1)
+			// this ca be caused by the client cache which is not up-to-date immediately after the first secret is created.
+			logger.Error(fmt.Errorf("invalid number of secrets found. actual %d, expected %d", len(secretList.Items), 1), "defaulting on the first secret in alphabetical order")
+			// Sort alphabetically the list of secrets
+			// so that in case of multiple secrets for the same namespace we always return the same one.
+			sort.Slice(secretList.Items, func(i, j int) bool {
+				return secretList.Items[i].Name < secretList.Items[j].Name
+			})
+			// let's default on the first secret, as they should all be valid, and they will all be deleted once the SpaceRequest resource is deleted.
+			namespaceAccess = append(namespaceAccess, toolchainv1alpha1.NamespaceAccess{
+				Name:      namespace.Name,
+				SecretRef: secretList.Items[0].Name,
+			})
```

Comment:
should we cleanup by deleting the other secrets? What if we just delete the other secrets and reconcile again?

Reply:
Yeah, it's a good idea indeed! 
But how do we know which secret we can delete?

I was thinking if it could happen that:
1) a first reconcile creates the first secret, let's say `secret-123` and sets the `SpaceRequest.Status.NamespaceAccess[0].SecretRef=secret-123`
 
2) meanwhile the "client" starts using this secret for accessing the namespace 

3) a second reconcile doesn't have this secret in the cache and creates a new secret `secret-456`, and updates the field to `SpaceRequest.Status.NamespaceAccess[0].SecretRef=secret-456`

4) the client might start using this new secret now, or it might still use the old one if it doesn't check constantly for changes on the SpaceRequest resource after it's provisioned 

how can we be sure about which secret to delete without creating "service interruption" ? 

Or should we just assume the client should "watch" after the SpaceRequest CR and always be using the secret that's present in the `SecretRef` field?

Reply:
We may implement the following piece of logic.

When creating the Secret (`case` at L446), if the field `SpaceRequest.Status.NamespaceAccess` contains an entry for the current namespace, use the entry's value for `SecretRef.Name` to as secret name, otherwise use the `GeneratedName` property for creating a new one.
This would recreate the same secret if it does not exist, or return an error if the secret already exists but we reached that place because the cache is not up to date.

When it comes to cleanup, we delete all the ones not in `SpaceRequest.Status.NamespaceAccess`.

WDYT?

Reply:
> Yeah, it's a good idea indeed! But how do we know which secret we can delete?

For selecting the ones we can delete, I'd say we should pick the ones not in `SpaceRequest.Status.NamespaceAccess`

> Or should we just assume the client should "watch" after the SpaceRequest CR and always be using the secret that's present in the SecretRef field?

Yeah, that's a safe approach that could be used by clients

Reply:
> This would recreate the same secret if it does not exist, or return an error if the secret already exists but we reached that place because the cache is not up to date.

How about we create the secret (case at L446) only when `SecretRef.Name` is empty ? we don't really need to create the secret with the same name and trigger an error right ?

> When it comes to cleanup, we delete all the ones not in SpaceRequest.Status.NamespaceAccess

I think we could cleanup only when `SpaceRequest.Status.NamespaceAccess` has already a `SecretRef.Name` for the specific namespace and the listing found more then one. is this what you are suggesting?

Reply:
> How about we create the secret (case at L446) only when `SecretRef.Name` is empty ? we don't really need to create the secret with the same name and trigger an error right ?

Then if the secret is manually deleted, we won't recreate that. Is this a scenario we want/have to manage?
Also, as the Secret is created in a loop and SpaceRequest's status is updated after the loop is complete, we may have an error in the middle of the loop and find ourself with status not correctly up to date. This will lead to more than one secret created. We can fix this by, catching the error, updating the status and rethrowing the error.

> I think we could cleanup only when `SpaceRequest.Status.NamespaceAccess` has already a `SecretRef.Name` for the specific namespace and the listing found more then one. is this what you are suggesting?

Reply:
@filariow thanks for your suggestion. 

Please see https://github.com/codeready-toolchain/host-operator/pull/953/commits/81e18f2c048465b55e1214cff27ff0d6727dc741 , where I'm checking for an existing `secretRef` and if not found, because of cache delay or secret doesn't exist, I'm now trying to recreate it with the same name, otherwise it creates a new secret using the `generateName` field.